backend/apps/webui/main.py

```diff
@@ -200,162 +196,165 @@ async def get_pipe_models():
-            function_module = app.state.FUNCTIONS[pipe_id]
+    mappings = {
+        "temperature": float,
+        "top_p": int,
+        "max_tokens": int,
+        "frequency_penalty": int,
+        "seed": lambda x: x,
+        "stop": lambda x: [bytes(s, "utf-8").decode("unicode_escape") for s in x],
+    }
 
-        if hasattr(function_module, "valves") and hasattr(function_module, "Valves"):
+    for key, cast_func in mappings.items():
+        if (value := params.get(key)) is not None:
+            form_data[key] = cast_func(value)
 
-            valves = Functions.get_function_valves_by_id(pipe_id)
-            function_module.valves = function_module.Valves(
-                **(valves if valves else {})
-            )
+    return form_data
 
-        pipe = function_module.pipe
 
-        # Get the signature of the function
-        sig = inspect.signature(pipe)
-        params = {"body": form_data}
+# inplace function: form_data is modified
+def populate_system_message(params: dict, form_data: dict, user) -> dict:
+    system = params.get("system", None)
+    if not system:
+        return form_data
 
-        if "__user__" in sig.parameters:
-            __user__ = {
-                "id": user.id,
-                "email": user.email,
-                "name": user.name,
-                "role": user.role,
-            }
+    if user:
+        template_params = {
+            "user_name": user.name,
+            "user_location": user.info.get("location") if user.info else None,
+        }
+    else:
+        template_params = {}
+    system = prompt_template(system, **template_params)
+    form_data["messages"] = add_or_update_system_message(
+        system, form_data.get("messages", [])
+    )
+    return form_data
 
-            try:
-                if hasattr(function_module, "UserValves"):
-                    __user__["valves"] = function_module.UserValves(
-                        **Functions.get_user_valves_by_id_and_user_id(pipe_id, user.id)
-                    )
-            except Exception as e:
-                print(e)
 
-            params = {**params, "__user__": __user__}
+async def generate_function_chat_completion(form_data, user):
+    model_id = form_data.get("model")
+    model_info = Models.get_model_by_id(model_id)
+    metadata = form_data.pop("metadata", None)
+
+    # Add extra params such as __event_emitter__
+    extra_params = get_extra_params(metadata)
+    if model_info:
+        if model_info.base_model_id:
+            form_data["model"] = model_info.base_model_id
 
-        if "__event_emitter__" in sig.parameters:
-            params = {**params, "__event_emitter__": __event_emitter__}
+        params = model_info.params.model_dump()
+        form_data = add_model_params(params, form_data)
+        form_data = populate_system_message(params, form_data, user)
 
-        if "__event_call__" in sig.parameters:
-            params = {**params, "__event_call__": __event_call__}
+    async def job():
```

Comment:
Is there a need for a separate `job()` function here? Or could this be flattened?

Reply:
Was used for sync def, but I believe It can be flattened now